File: apps/ios/Shared/Views/LocalAuth/SetAppPasscodeView.swift
//
//  SetAppPaswordView.swift
//  SimpleX (iOS)
//
//  Created by Evgeny on 10/04/2023.
//  Copyright © 2023 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

struct SetAppPasscodeView: View {
    var passcodeKeychain: KeyChainItem = kcAppPassword
    var prohibitedPasscodeKeychain: KeyChainItem = kcSelfDestructPassword
    var title: LocalizedStringKey = "New Passcode"
    var reason: String?
    var submit: () -> Void
    var cancel: () -> Void
    @Environment(\.dismiss) var dismiss: DismissAction
    @State private var showKeychainError = false
    @State private var passcode = ""
    @State private var enteredPassword = ""
    @State private var confirming = false

    var body: some View {
        ZStack {
            if confirming {
                setPasswordView(
                    title: "Confirm Passcode",
                    submitLabel: "Confirm",
                    submitEnabled: { pwd in pwd == enteredPassword }
                ) {
                    if passcode == enteredPassword {
                        if passcodeKeychain.set(passcode) {
                            enteredPassword = ""
                            passcode = ""
                            dismiss()
                            submit()
                        } else {
                            showKeychainError = true
                        }
                    }
                }
            } else {
                setPasswordView(title: title,
                                submitLabel: "Save",
                                // Do not allow to set app passcode == selfDestruct passcode
                                submitEnabled: { pwd in pwd != prohibitedPasscodeKeychain.get() }) {
                    enteredPassword = passcode
                    passcode = ""
                    confirming = true
                }
            }
        }
        .alert(isPresented: $showKeychainError) {
            mkAlert(title: "KeyChain error", message: "Error saving passcode")
        }
    }

    private func setPasswordView(title: LocalizedStringKey, submitLabel: LocalizedStringKey, submitEnabled: (((String) -> Bool))? = nil, submit: @escaping () -> Void) -> some View {
        PasscodeView(passcode: $passcode, title: title, reason: reason, submitLabel: submitLabel, submitEnabled: submitEnabled, buttonsEnabled: Binding.constant(true), submit: submit) {
            dismiss()
            cancel()
        }
    }
}

struct SetAppPasscodeView_Previews: PreviewProvider {
    static var previews: some View {
        SetAppPasscodeView(submit: {}, cancel: {})
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/localauth/SetAppPasscodeView.kt
package chat.simplex.common.views.localauth

import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import chat.simplex.common.platform.BackHandler
import chat.simplex.common.views.helpers.DatabaseUtils
import chat.simplex.common.views.helpers.DatabaseUtils.ksAppPassword
import chat.simplex.common.views.helpers.DatabaseUtils.ksSelfDestructPassword
import chat.simplex.common.views.helpers.generalGetString
import chat.simplex.res.MR

@Composable
fun SetAppPasscodeView(
  passcodeKeychain: DatabaseUtils.KeyStoreItem = ksAppPassword,
  prohibitedPasscodeKeychain: DatabaseUtils.KeyStoreItem = ksSelfDestructPassword,
  title: String = generalGetString(MR.strings.new_passcode),
  reason: String? = null,
  submit: () -> Unit,
  cancel: () -> Unit,
  close: () -> Unit
) {
  val passcode = rememberSaveable { mutableStateOf("") }
  var enteredPassword by rememberSaveable { mutableStateOf("") }
  var confirming by rememberSaveable { mutableStateOf(false) }

  @Composable
  fun SetPasswordView(title: String, submitLabel: String, submitEnabled: (((String) -> Boolean))? = null, submit: () -> Unit) {
    BackHandler {
      close()
      cancel()
    }
    PasscodeView(passcode, title = title, reason = reason, submitLabel = submitLabel, submitEnabled = submitEnabled, submit = submit) {
      close()
      cancel()
    }
  }

  if (confirming) {
    SetPasswordView(
      generalGetString(MR.strings.confirm_passcode),
      generalGetString(MR.strings.confirm_verb),
      submitEnabled = { pwd -> pwd == enteredPassword }
    ) {
      if (passcode.value == enteredPassword) {
        passcodeKeychain.set(passcode.value)
        enteredPassword = ""
        passcode.value = ""
        close()
        submit()
      }
    }
  } else {
    SetPasswordView(title, generalGetString(MR.strings.save_verb),
      // Do not allow to set app passcode == selfDestruct passcode
      submitEnabled = { pwd -> pwd != prohibitedPasscodeKeychain.get() }) {
      enteredPassword = passcode.value
      passcode.value = ""
      confirming = true
    }
  }
}
